File: day25.lua
#!/usr/bin/env lua

local input = io.open("day25.txt"):read("a")
local locks = {}
local keys = {}

for grid in input:gsub("\n\n", "\0"):gmatch("[^\0]+") do
    local levels = {}
    local kind
    local i = 1
    for row in grid:gmatch("[^\n]+") do
        local j = 1
        for cell in row:gmatch(".") do
            kind = kind or cell
            if cell == kind then levels[j] = i end
            j = j + 1
        end
        i = i + 1
    end
    table.insert(kind == "#" and locks or keys, levels)
end

local function fits(key, lock)
    for i = 1, #lock do
        if lock[i] > key[i] then return false end
    end
    return true
end

local count = 0
for _, key in ipairs(keys) do
    for _, lock in ipairs(locks) do
        if fits(key, lock) then count = count + 1 end
    end
end
print(count)

File: day25.c
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define N 500
#define W 5
int locks[N][W];
int nlocks = 0;
int keys[N][W];
int nkeys = 0;

bool fits(int* key, int* lock)
{
    for (int i = 0; i < W; i++) {
        if (lock[i] > key[i]) return false;
    }
    return true;
}

int main()
{
    FILE* file = fopen("day25.txt", "r");

    for (char line[W + 2]; fgets(line, sizeof(line), file); ) {
        int levels[W] = { 0 };
        char kind = 0;
        for (int i = 0; strlen(line) >= W; i++) {
            for (int j = 0; j < W; j++) {
                char cell = line[j];
                if (!kind) kind = cell;
                if (cell == kind) levels[j] = i;
            }
            if (!fgets(line, sizeof(line), file)) break;
        }
        memcpy(kind == '#' ? locks[nlocks++] : keys[nkeys++],
            levels, W * sizeof(levels[0]));
    }

    int count = 0;
    for (int i = 0; i < nkeys; i++) {
        for (int j = 0; j < nlocks; j++) {
            if (fits(keys[i], locks[j])) count++;
        }
    }
    printf("%d\n", count);
}
